### PR TITLE
Fix errors in LB+Lees-Edwards

### PR DESCRIPTION
**Pull request by RudolfWeeber**
contains #4827

This 
* fixs index calculation in the LB Walberla sweep doing the interpolation of populations at the Lees-Edwards boundary
* adds missing Lees-Edwards velocity shift in the lb particle coupling, when a couplign is poerformed at a particle positoin +- box_length (for ghosts)
* adds more testing of LE+LB particle coupling

- LB+Lees Edwards fixes
- Allow Langevin +LB
- LB: Fix rounding error in particle coupling
- LB+particles: guard against missed coupling due to round-off error
- Formatting

Fixes #

Description of changes:
-

### CODE REVIEW

File: src/core/lb/particle_coupling.cpp [hunk 1]
```diff
@@ -186,10 +186,17 @@ std::vector<Utils::Vector3d> positions_in_halo(Utils::Vector3d const &pos,
   auto const halo = 0.5 * agrid;
   auto const halo_vec = Utils::Vector3d::broadcast(halo);
   auto const fully_inside_lower = local_geo.my_left() + 2. * halo_vec;
+
+  // If the particle is at least one agrid away from the node boundary
+  // any ghosts shifte dby +- box_length cannot be in the lb volume
```
Comment (christophlohrmann) [inline src/core/lb/particle_coupling.cpp]: typo

File: src/core/lb/particle_coupling.cpp [hunk 2]
```diff
@@ -211,14 +218,35 @@ std::vector<Utils::Vector3d> positions_in_halo(Utils::Vector3d const &pos,
...
+  if (box_geo.type() == BoxType::LEES_EDWARDS) {
+    auto le = box_geo.lees_edwards_bc();
+    auto normal_shift = (shifted_pos - orig_pos)[le.shear_plane_normal];
+    if (normal_shift > std::numeric_limits<double>::epsilon()) {
+      vel_shift[le.shear_direction] -= le.shear_velocity;
+    }
+    if (normal_shift < -std::numeric_limits<double>::epsilon()) {
+      vel_shift[le.shear_direction] += le.shear_velocity;
+    }
+  }
```
Comment (christophlohrmann) [inline src/core/lb/particle_coupling.cpp:189]: is it on purpose that nothing happens if the normal shift is inbetween -epsilon and epsilon?

File: src/walberla_bridge/src/lattice_boltzmann/InterpolateAndShiftAtBoundary.hpp [hunk 3]
```diff
@@ -111,24 +119,30 @@ class InterpolateAndShiftAtBoundary {
...
       Cell source2 = cell;
-      source1[dir] = cell_idx_c(std::floor(
-                         static_cast<FloatType>(source1[dir]) + offset)) %
-                     dim;
-      source1[dir] = cell_idx_c(static_cast<FloatType>(source1[dir]) + length);
-      source1[dir] = cell_idx_c(source1[dir] % dim);
-
+      int target_idx = cell[dir];
+      auto ghost_fold = [dim](FloatType x) {
+        while (x > dim) {
```
Comment (christophlohrmann) [inline src/walberla_bridge/src/lattice_boltzmann/InterpolateAndShiftAtBoundary.hpp]: this also looks a lot like a modulo

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 4]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
+        # LB Nodes surrounding the particle.
+        # On the particle's side 
+        unshifted = [lbf[mid_x - 1, 0, mid_z - 1], 
+                     lbf[mid_x, 0, mid_z - 1],
+                     lbf[mid_x - 1, 0, mid_z],
+                     lbf[mid_x, 0, mid_z]]
+        # across the Lees Edwads boundary conditoin
+        # if offset ^1<.5, it couples to the left neighbor
+        # otherwise to the right
+        if (offset % 1) >= .5: extra = 1
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: ternary operator would look nicer than if-else without indentation

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 5]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
-                 lbf[mid_x + idx, upper_y, mid_z]]
+        upper_y = lbf.shape[1] - 1  # y is shear plane noremal
+        # LB Nodes surrounding the particle.
+        # On the particle's side 
+        unshifted = [lbf[mid_x - 1, 0, mid_z - 1], 
+                     lbf[mid_x, 0, mid_z - 1],
+                     lbf[mid_x - 1, 0, mid_z],
+                     lbf[mid_x, 0, mid_z]]
+        # across the Lees Edwads boundary conditoin
+        # if offset ^1<.5, it couples to the left neighbor
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: typo?

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 6]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
+            np.testing.assert_allclose(
+                np.copy(n.last_applied_force), -np.copy(p.f) * 1 / 2 * 1 / 2 * weight_l)
+        for n in shifted_right:
+            np.testing.assert_allclose(
+                np.copy(n.last_applied_force), -np.copy(p.f) * 1 / 2 * 1 / 2 * weight_r)
 
+        # Check the LB velocity interpolation
+        # at a positoin, where the ghost layer of the
+        # lees edwards shear plane contributes
+        lbf[:, :, :].velocity = np.zeros(3)
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: duplication

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 7]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
-        lower_nodes = nodes[:4]
-        upper_nodes = nodes[4:]
+        lower_nodes = unshifted
+        upper_nodes = shifted_left + shifted_right
         for n in lower_nodes:
             n.velocity = v0
         for n in upper_nodes:
             n.velocity = - v0
+        # When the offset modulo 1 is not zero
+        # in the ghost layer, neighboring cell swith zero velocity contribute
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: typo

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 8]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
-        np.testing.assert_allclose(np.copy(p.pos), pos)
-        np.testing.assert_allclose(np.copy(p.f), np.zeros(3))
-        for n in nodes:
-            np.testing.assert_allclose(
-                np.copy(n.last_applied_force), np.zeros(3))
+            expected_v)
+
+    def test_viscous_coupling_with_shear_vel(self):
+        # Places a co-moving particle close to the LE boundary
+        # in shear flow. chesk that it remains force free
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: typo

File: testsuite/python/lb_lees_edwards_particle_coupling.py [hunk 9]
```diff
@@ -45,49 +45,119 @@ def test(self):
...
+        # this is only the case, if the periodic imagesin the
```
Comment (christophlohrmann) [inline testsuite/python/lb_lees_edwards_particle_coupling.py]: typo

**Review by christophlohrmann — COMMENTED** [review]
looks good but I'm no expert in lees edwards

File: src/walberla_bridge/src/lattice_boltzmann/InterpolateAndShiftAtBoundary.hpp (hunk 3 above)
Comment (RudolfWeeber) [reply src/walberla_bridge/src/lattice_boltzmann/InterpolateAndShiftAtBoundary.hpp]: It isn't, actually. A modulo would be x>=n or x<n. Here, it's x>n or x<-1. This is, ebcause Walberla uses -1 and n to address the ghost layers.

File: src/core/lb/particle_coupling.cpp (hunk 2 above)
Comment (RudolfWeeber) [reply src/core/lb/particle_coupling.cpp:189]: Yes. normal_shift contains multiples of box_l or 0. We basally just want +-1 or 0, but tollerant to rounding errors. However, in one of teh two instances, where this appears, that integer is directly available in the shift variable, so I used that instead. In the other instance, I tried to make it clear by renaming the variables.

File: testsuite/python/lb_lees_edwards_particle_coupling.py (hunk 4 above)
Comment (RudolfWeeber) [reply testsuite/python/lb_lees_edwards_particle_coupling.py]: Replaced by round()

**Comment by RudolfWeeber** [comment]
I'm closing this in favor of a new pr

**Pull request closed without merging**